compiler/rustc_const_eval/src/interpret/operand.rs

```diff
@@ -243,6 +243,17 @@ impl<'tcx, Prov: Provenance> std::ops::Deref for ImmTy<'tcx, Prov> {
 }
 
 impl<'tcx, Prov: Provenance> ImmTy<'tcx, Prov> {
+    #[inline(always)]
+    pub fn try_from_immediate(imm: Immediate<Prov>, layout: TyAndLayout<'tcx>) -> Option<Self> {
```

Comment:
I'm not a fan of these functions. The point of the assertions is that they are a last line of defense to detect defective callers. They are *not* exhaustive checks. If the caller can't ensure that the value has the right type, that can only be fixed in the caller.

IOW, `matches_abi` here really is more of a `maybe_matches_abi`. It is necessary, but not sufficient. Ad trying to make it sufficient is the wrong approach; the right approach is figuring out why someone is feeding bogus data into these functions.